### ISSUE TITLE
What are Session.SetSessionDestroyCallback and Session.GetDestroyCallback?

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#2722]
Starcounter.Session has two methods:

1. `Session.SetSessionDestroyCallback`
2. `Session.GetDestroyCallback`

Are they getting/setting the same value? If so, then why the names are different?

What are those methonds intend to do? If this is some kind of event, which occurs when session dies, then why not use `event` instead of `Action`, so any app may subscribe and don't worry about overwriting value from other app?

@alemoi, @k-rus, @chrhol I guess some of you can answer me.

**Comment by alemoi** [comment]
Good question!
There several reasons the session can die. One of those is timeout (set in `DefaultSessionTimeoutMinutes`) in database config.

`GetDestroyCallback` just returns what one sets in `SetSessionDestroyCallback`.

> What are those methonds intend to do? If this is some kind of event, which occurs when session dies, > then why not use event instead of Action, so any app may subscribe and don't worry about overwriting > value from other app?

That's a good suggestion that we should consider. Maybe they should work as middleware filters. We can discuss that!

**Comment by miyconst** [comment]
Does not the session dies after user closes all browser windows? I am trying to use this event to delete uploaded, and not saved files, but unable to find how to trigger this action.

Refresh page, close browser, stop app, stop server, does not trigger this action.

**Comment by alemoi** [comment]
> Refresh page, close browser, stop app, stop server, does not trigger this action.

This is correct since server does not know if it should kill the session or not.

Regarding WebSockets they are now decoupled from sessions: when WebSocket connection closes, it does not destroy the session. But I can fix it back if this behavior is needed.

Regarding connection-less protocols, the client should probably do something (like call a specific REST handler) in order to kill the session.

**Comment by miyconst** [comment]
No, I think the session should live even if the socket connection died. In theory, the SessionEnd event may not fire at all, for example if server lost power. That is why I ended up deleting old files on application startup.

Anyway, if `Starcounter.Session` has a `SessionDestoyed` event, it should be an `event`, not an `Action`.

**Comment by alemoi** [comment]
The session destruction API has now changed to usage of standard `EventHandler` property on `Session`: `public event EventHandler Destroyed`.

Fixed via acb6dc07818ebbbc6e9e0d200d82035dd2ee552d 

The following test demonstrates the new interface:

```c#
Session s = new Session();

String testString = "";

// Setting the destroyed events.
s.Destroyed += (object sender, EventArgs e) => {
    Assert.AreEqual(s, (Session)sender);
    testString += "One";
};
s.Destroyed += (object sender, EventArgs e) => {
    Assert.AreEqual(s, (Session)sender);
    testString += "Two";
};

// Destroying the session.
s.Destroy();
Assert.AreEqual("OneTwo", testString);
```
 
Thanks for the input!

**Comment by miyconst** [comment]
Cool!
